### ISSUE TITLE
InputRadioGroup not registering initial click in some circumstances

### ISSUE THREAD

**Issue by TimCorey** [dotnet/aspnetcore#39145]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When I have made an edit to an input field, if I click a button afterward, the button click does not fire.
 
This seems similar to #30875 but the solution is not the same. In the simple demo, wrapping the string in a model does seem to resolve the issue. However, in my production issue, I have a radio button that cannot be bound directly to the model. It needs to be translated first. It exhibits the same behavior, where clicking on an option the first time does not select it if I have first made an edit to an input field. The second click works.

### Expected Behavior

When I edit a text file and then click (using the mouse) on the next field or on a button, that click should be registered.

### Steps To Reproduce

Here are the repo steps:
1.	Go to [https://blazorformsample.azurewebsites.net/](https://blazorformsample.azurewebsites.net/)
2.	Click the submit button. Notice the counter increments.
3.	Modify the text in the text field.
4.	Click (don’t tab or hit enter) on the Submit button. Nothing happens.
5.	Click the button again and it fires.

Here is the sample repo: [https://github.com/TimCorey/BlazorFormSample](https://github.com/TimCorey/BlazorFormSample)

### .NET Version

6.0.101

### Anything else?

I have reproduced this on Mac and Windows 10, on FireFox, Chrome, and Edge. Some people seem to not experience it at first, but then will encounter it after a few attempts.

Here is a demo showing off the bug:
![2021-12-15_16-36-29](https://user-images.githubusercontent.com/1839873/147006330-c820cab7-e0d8-4aa6-af7c-118877bb28e9.gif)
.

**Comment by MariovanZeist** [comment]
The underlying cause is the use of a value type as ``Model``.
The ``EditForm`` uses the ``HashCode`` of the ``Model`` to determine if anything inside the ``EditForm`` should be recreated.
So if the ``Model`` changes, it will recreate its contents (including the ``<form>`` tag and the ``onsubmit`` attribute).
As the ``Model`` is of type ``object``, the integer ``@currentCount`` gets boxed, this boxing happens every time the ``EditForm`` is rendered.
This will result in a different ``HashCode`` for every render, and in turn, will make ``EditForm`` recreate its ``ChildContent`` every render as well.

You can see this by creating a simple component like this:

Construct.razor
```
<h3>I am constructed:@_counter times</h3>

@code {
    private static int _staticCounter=0;
    private int _counter = ++_staticCounter;
}
```

and add that below your button.

You will see it recreates the component every time you change anything inside the ``EditForm``. (Change text, press button)
If you change the ``Model`` to `` this`` that disappears, and it will create it only once. (It will probably show 2 because of the prerendering)

So in your example when you change the text, and then without pressing enter or tab, click on the submit button,
As soon as you leave the input, the ``OnChange`` of that ``Input`` will fire, which in turn will call ``StateHasChanged`` on the ``EditForm``, which in turn will rerender the ``EditForm``, and because the boxed int having a different ``HashCode`` it will also in turn recreate
the entire ``ChildContent`` (including the ``<form>`` and its ``onsubmit`` attribute to be recreated, resulting in a race condition which explains why some people don't see it, and others do)

So my recommendation would be to always use a reference type as a ``Model``, not only for your submit problem but also for the needless recreating of Components.

**Comment by MariovanZeist** [comment]
I do think this should be better documented, either in the EditForms Model Intellisense, or 
https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-6.0

**Comment by TimCorey** [comment]
Thanks @MariovanZeist that was really helpful. I might add the documentation necessary to that section because I agree that it should be clearer.

However, I am still having the issue even after moving my radio button option into my model (I created a model just for this form). Here is my actual code:

`
<EditForm Model="suggestion" OnValidSubmit="Create">
    <DataAnnotationsValidator />
    <ValidationSummary />
    <div class="input-section">
        <InputText id="suggestion-text" @bind-Value="suggestion.Suggestion" class="form-control" />
    </div>

    <div class="input-section">
            <InputRadioGroup id="category" @bind-Value="suggestion.CategoryId">
                @foreach (var c in categories)
                {
                <div class="radio-item-group">
                    <InputRadio Value="@c.CategoryId" id="@c.CategoryId" />
                    <label for="@c.CategoryId">@c.CategoryName - @c.CategoryDescription</label>
                </div>
                }
            </InputRadioGroup>
    </div>
    <div class="input-section">
        <InputTextArea id="description" @bind-Value="suggestion.Description" class="form-control" />
    </div>
    <div class="center-children">
        <button class="btn btn-main btn-lg text-uppercase" type="submit">Suggest</button>
    </div>
</EditForm>
`

The model items are just three string values. When I enter text into either Suggestion or Description, I cannot select the radio option without a second click. I don't show the actual radio items, just their labels but I don't think that should be an issue.

**Comment by MariovanZeist** [comment]
Hi @TimCorey 
This is a similar issue, now caused by the ``InputRadioGroup`` component. 
Basically, it will always rebuild its internals (So in your case the ``InputRadio`` components) due to an always-changing Key, resulting in the same kind of race condition as above.

This seems to be a duplicate of: https://github.com/dotnet/aspnetcore/issues/31610

**Comment by MariovanZeist** [comment]
As a workaround you could create your own component  as replacement for ``InputRadioGroup`` like this:

```cs
 public class MyInputRadioGroup<TValue> : InputRadioGroup<TValue>
 {
     private string? _name;
     private string? _fieldClass;

     protected override void OnParametersSet()
     {
         var fieldClass = EditContext?.FieldCssClass(FieldIdentifier) ?? string.Empty;
         if (fieldClass != _fieldClass || Name != _name)
         {
             _fieldClass = fieldClass;
             _name = Name;
             base.OnParametersSet();
         }
     }
 }
 ```

**Comment by TimCorey** [comment]
Thanks @MariovanZeist, that was very helpful and allowed me to at least get a workaround in place for my issue.

I'd be interested to know why this isn't the default behavior of the InputRadioGroup.

**Comment by zarinportal** [comment]
I have exactly the same issue, I must double-click to be selected


`	<InputRadioGroup @bind-Value="request.RentOptionId">
				@foreach (var option in options)
				{
					<label for="option@(option.Id)"><InputRadio Value="option.Id" id="@("option" + option.Id)" /> @option.NameEn</label>
				}
		</InputRadioGroup>`

**Comment by zarinportal** [comment]
i created class with InputRadioGroupWithForeach name and use it, like this:



```
public class InputRadioGroupWithForeach<TValue> : InputRadioGroup<TValue>
    {
        private string? _name;
        private string? _fieldClass;

        protected override void OnParametersSet()
        {
            var fieldClass = EditContext?.FieldCssClass(FieldIdentifier) ?? string.Empty;
            if (fieldClass != _fieldClass || Name != _name)
            {
                _fieldClass = fieldClass;
                _name = Name;
                base.OnParametersSet();
            }
        }
    }
```

**Comment by SteveSandersonMS** [comment]
The original issue reported here doesn't involve radio groups, so although the PR https://github.com/dotnet/aspnetcore/pull/39208 may fix another issue, but I don't see how it can fix this one.

Anyone know how to persuade GitHub to unlink this issue from that PR? The UI seems to be disabled:

![image](https://user-images.githubusercontent.com/1101362/148767787-d7b62554-7648-45f5-a329-67fa049608b0.png)

**Comment by SteveSandersonMS** [comment]
Also, if the issue with radio groups has a similar underlying cause to the one being reported here, that may suggest we should look for a more general way to address it that isn't specific to radio groups.

(However if I'm misunderstanding the relationship between these issues I apologise!)

**Comment by MariovanZeist** [comment]
The 2 issues described here are unrelated. (The only relation is that they both have to do with user Input)
The first issue had to do with the fact that the ``Model`` value of an ``EditForm`` was set to a value type, causing the ``EditForm`` to recreate its ``ChildContent`` every time the ``EditForm`` rerenders (For example due to a change in an input).

The second issue stated in the above comment  https://github.com/dotnet/aspnetcore/issues/39145#issuecomment-999744378  is specific to the ``InputRadioGroup`` component.

**Comment by SteveSandersonMS** [comment]
Fixed in https://github.com/dotnet/aspnetcore/pull/40148
